ppd/testppdfile.c

```diff
@@ -0,0 +1,179 @@
+  int files;			   /* Number of files */
+  cups_array_t	*output;
+  int len_output;  /* Length of the output array */
+  char[256] txt;
+  int warn;			   /* Which errors to just warn about */
+  int ignore;			   /* Which errors to ignore */
+
+
+  verbose = 0;
+  root_present = 0;
+  help = 0;
+  relaxed = 0;
+  q_with_v = 0;
+  v_with_q = 0;
+  files=0;
+  warn = WARN_NONE;
+  ignore = WARN_NONE;
+
+  for (i = 1; i < argc; i++)
+    if (!strcmp(argv[i], "--help"))
+    help = 1;
+  else if (argv[i][0] == '-' && argv[i][1])
+  {
+    for (opt = argv[i] + 1; *opt; opt++)
+      switch (*opt)
+      {
+        case 'I': /* ignore_params errors */
+            i++;
+
+            if (i >= argc)
+              help = 1;
+
+            if (!strcmp(argv[i], "none"))
+              ignore = WARN_NONE;
+	        else if (!strcmp(argv[i], "filename"))
+	          ignore |= WARN_FILENAME;
+	        else if (!strcmp(argv[i], "filters"))
+  	          ignore |= WARN_FILTERS;
+	        else if (!strcmp(argv[i], "profiles"))
+	          ignore |= WARN_PROFILES;
+	        else if (!strcmp(argv[i], "all"))
+	          ignore = WARN_FILTERS | WARN_PROFILES;
+            else
+              help = 1;
+            break;
+
+        case 'R': /* Alternate root directory */
+            i++;
+
+            if (i >= argc)
+              help = 1;
+
+              rootdir = argv[i];
+              root_present = 1;
+              break;
+
+        case 'W': /* Turn errors into warn_paramsings */
+            i++;
+
+            if (i >= argc)
+              help = 1;
+
+            if (!strcmp(argv[i], "none"))
+              warn = WARN_NONE;
+	        else if (!strcmp(argv[i], "constraints"))
+	          warn |= WARN_CONSTRAINTS;
+	        else if (!strcmp(argv[i], "defaults"))
+	          warn |= WARN_DEFAULTS;
+	        else if (!strcmp(argv[i], "duplex"))
+	          warn |= WARN_DUPLEX;
+	        else if (!strcmp(argv[i], "filters"))
+	          warn |= WARN_FILTERS;
+	        else if (!strcmp(argv[i], "profiles"))
+	          warn |= WARN_PROFILES;
+	        else if (!strcmp(argv[i], "sizes"))
+	          warn |= WARN_SIZES;
+	        else if (!strcmp(argv[i], "translations"))
+	          warn |= WARN_TRANSLATIONS;
+	        else if (!strcmp(argv[i], "all"))
+	          warn = WARN_ALL;
+            else
+              help = 1;
+            break;
+
+        case 'q': /* Quiet mode */
+
+            if (verbose > 0)
+            {
+              q_with_v = 1;
+            }
+            verbose--;
+            break;
+
+        case 'r': /* Relaxed mode */
+            relaxed = 1
+
+        case 'v': /* Verbose mode */
+            if (verbose < 0)
+            {
+              v_with_q = 1
+            }
+
+            verbose++;
+            break;
+
+        default:
+            help = 1;
+      }
+  }
+  else
+  {			
+    files++;
+
+    if (argv[i][0] == '-')
+    {
+      _ppdArrayAddStrings(file_array,"");
```

Comment:
For these simple additions you could use `cupsArrayNew()` and `cupsArrayAdd()` instead off the functions of libppd's private API, to keep the private API inside libppd. Then you could also remove `#include <ppd/array-private.h>` at the top of the file.